webapp/tests/Utils/Scoreboard/ScoreboardIntegrationTest.php

```diff
@@ -0,0 +1,281 @@
+            ->setStarttimeString(date('Y').'-01-01 10:00:00 Europe/Amsterdam')
+            ->setActivatetimeString('-01:00')
+            ->setEndtimeString('+02:00');
+        $this->em->persist($this->contest);
+
+        $hostname = self::CONTEST_NAME.'-judgehost';
+        $this->judgehost = $this->em->getRepository(Judgehost::class)
+            ->findOneBy(['hostname' => $hostname]);
+        if ( !$this->judgehost ) {
+            $this->judgehost = new Judgehost();
+            $this->judgehost
+                ->setHostname($hostname);
+            $this->em->persist($this->judgehost);
+        }
+
+        $category = $this->em->getRepository(TeamCategory::class)
+            ->findOneBy(['sortorder' => 0]);
+
+        for($i=0; $i<self::NUM_TEAMS; $i++) {
+            $this->teams[$i] = new Team();
+            $this->teams[$i]
+                ->setName('Team '.$i)
+                ->setCategory($category);
+            $this->em->persist($this->teams[$i]);
+       }
+
+        for($i=0; $i<self::NUM_PROBLEMS; $i++) {
+            $letter = chr(ord('a') + $i);
+            $this->problems[$i] = new Problem();
+            $this->problems[$i]
+                ->setName('Problem '.$letter)
+                ->setTimelimit(5);
+
+            $cp = new ContestProblem();
+            $cp
+                ->setShortname($letter)
+                ->setProblem($this->problems[$i])
+                ->setContest($this->contest);
+
+            $this->contest->addProblem($cp);
+            $this->problems[$i]->addContestProblem($cp);
+
+            $this->em->persist($this->problems[$i]);
+            $this->em->persist($cp);
+        }
+
+        $this->em->flush();
+    }
+
+    protected function tearDown()
+    {
+        // Preserve the data for inspection if a test failed.
+        if ( !$this->hasFailed() ) {
+            $this->em->remove($this->contest);
+            $this->em->remove($this->judgehost);
+            foreach ($this->teams    as $team)    $this->em->remove($team);
+            foreach ($this->problems as $problem) $this->em->remove($problem);
+        }
+
+        $this->em->flush();
+
+        parent::tearDown();
+
+        $this->em->close();
+        $this->em = null; // avoid memory leaks
+    }
+
+    public function testScoreboardsNoFreeze()
+    {
+        $this->contest->setFreezetimeString(null);
+        $this->createDefaultSubmissions();
+        $this->recalcScoreCaches();
+
+        $expected_scores = [
+            [ 'team' => $this->teams[0], 'rank' => 2, 'solved' => 0, 'time' => 0 ],
+            [ 'team' => $this->teams[1], 'rank' => 1, 'solved' => 2, 'time' => 161 ],
+        ];
+
+        foreach ([ false, true ] as $jury) {
+            $scoreboard = $this->ss->getScoreboard($this->contest, $jury);
+            $this->assertScoresMatch($expected_scores, $scoreboard->getScores());
+        }
+    }
+
+    public function testScoreboardJuryFreeze()
+    {
+        $this->createDefaultSubmissions();
```

Comment:
Why is there no `$this-> recalcScoreCaches()` here but there is one in `testScoreboardsNoFreeze`?

Reply:
There is, but it's in the for loop below that changes the freeze time. I'll add a comment to clarify.